### ISSUE TITLE
Add TreeTraverser.xxxTraversal(Iterable) overloads for multiple root nodes

### ISSUE THREAD

**Issue by bjmi** [google/guava#2009]
From time to time I have to traverse tree-based data structures where starting point are multiple root nodes  e.g. traversing files with TreeTraverser.breadthFirstTraversal(dir1, dir2, dir3) with one single call. 

One could sometimes create a single virtual root node and skip this node but it feels like a work-around.

**Comment by andrey-polyakov** [comment]
I am not quite sure what is your use case, and how to reproduce the issue.

Would you please put the problem in plain English with an example.

**Comment by bjmi** [comment]
Let's say you have a tree-based data structure like UI components where each component can be a container of children components.
My use case would be to travers several independent trees uniformly
`List<UIComponent> uiComponents = Lists.newArrayList(page1, page3, page5);`
e.g. in breadth first traversal
```
Optional<UICompontent> greenButton =
    treeTraverser.breadthFirstTraversal(uiComponents)
                 .firstMatch(greenButtonPredicate)
```
or like my directory traversal example
`Iterable<File> logFiles = TreeTraversers.breadthFirstTraversal(apacheLogDir, mysqlLogDir, ...);`

I hope this use case is clear now and not pure invention. Is there a better approach I overlooked?

**Comment by andrey-polyakov** [comment]
Right. It is a better overview but where exactly the problem strikes? When that odd thing happens? Thanks.

**Comment by bjmi** [comment]
Current TreeTraverser methods breadthFirstTraversal(T), postOrderTraversal(T) and preOrderTraversal(T) require a **single** root node.
I need something like breadthFirstTraversal(Iterable<T> roots), postOrderTraversal(Iterable<T> roots) or preOrderTraversal(Iterable<T> roots).

**Comment by ronshapiro** [comment]
@nymanjens @jrtom FYI^

**Comment by jrtom** [comment]
I'm not sure what it is that you expect the behavior of this proposed method to be.  I would guess that you would expect the output `Iterable` to be the concatenation of the independent traversal  starting from each of the supplied root nodes.

In that case you could just use `Iterables.concat()` to do what you want:

```
Traverser t = Traverser.forTree(n -> yourChildrenMethod(n));
Iterable nodeList = Iterables.concat(
    t.breadthFirst(page1), 
    t.breadthFirst(page3), 
    t.breadthFirst(page5)); 
```

(This uses the new `Traverser` class-currently in development but the principle is the same.)

Am I missing something, or would that do what you want?

I am somewhat reluctant to have a method on `Traverser` that allows you to specify multiple starting nodes, because the behavior is not intrinsically well-defined: what do you do, for example, if you encounter the same node starting from different start nodes?  How do you define what legal inputs are?

**Comment by jrtom** [comment]
Since this issue hasn't (until today) been updated for over a year, my take on this is:
(a) it doesn't sound like there's a lot of demand for this, and 
(b) the above solution (use `Iterables.concat()`) is what people should be doing, unless someone comes back and explains why that doesn't work for them.

So I'm closing this; if I've overlooked something or if there are enough people for which this mechanism is overly burdensome, we'll reopen it.

**Comment by bjmi** [comment]
The concat(...) approach provides expected order for pre-order iteration.
```
Traverser t = Traverser.forTree(n -> yourChildrenMethod(n));
Iterable nodeList = Iterables.concat(
    t.depthFirstPreOrder(page1), 
    t.depthFirstPreOrder(page3), 
    t.depthFirstPreOrder(page5));
```
Given breadthFirst(...) example would iterate as follows:
1. page1 and children in bf-order
2. page3 and children in bf-order
3. page5 and children in bf-order

but it was requested:
1. page1, page3 and page5
2. 1. level of page1, 1. level of page3 and 1. level of page5
3. 2. level of page1, 2. level of page3 and 2. level of page5
4. and so on

I'm still waiting for Guava implementation in order to replace my own implementation.

**Comment by jrtom** [comment]
Thanks for the clarification as to what you wanted, although the question of what such a method should do if it encounters the same node multiple times (which it certainly could, in general) is still open.

In order to do the specific thing that you want, I think that your first instinct--to use a virtual root node that has your starting points as successors--is actually the right one to use.  It's the right semantics for what you're trying to do, i.e., treat the union of the _k_ input trees (presuming that they are actually disjoint) as if they were subtrees of an implicit root.  You'd want to use `Iterables.skip()` to get past the virtual root, but that's easy enough:

```
SuccessorsFunction<N> successorsFunction =
    n -> { n == virtualRoot ? ImmutableList.of(page1, page3, page5) : yourChildrenMethod(n) };
Iterable<N> traversal = Iterables.skip(
    Traverser.forTree(successorsFunction).breadthFirst(virtualRoot),
    1);
```

IMO this doesn't sound like something that is likely to be of sufficiently broad interest to be a suitable candidate for inclusion in Guava.  And I think that it would be quite difficult to come up with a name that would correctly convey the semantics, because the semantics you requested are definitely not what I  expected.

I'm happy to reconsider this if one of the other Guava owners thinks that it's likely to be more broadly useful (and can propose a name that will convey the semantics you want).

**Comment by bjmi** [comment]
Thank you for implementing traversal of multiple root notes in [release 24.1](https://github.com/google/guava/releases/tag/v24.1)

> `graph:` Added support for traversing from multiple start nodes in `Traverser`. ([`cbc5308`](https://github.com/google/guava/commit/cbc53084465d3a95cdb8cbd79da21624e3086d5d))

**Comment by jrtom** [comment]
Yep, we gave it some additional thought and decided that it was worth doing.  Thanks for your patience.  :)
